message_ix/macro.py

```diff
@@ -1,61 +1,464 @@
+            'idx': ['node', 'year', ],
+            'unit': 'G$',
+            'data_key': 'total_cost',
+        },
+        'gdp_calibrate': {
+            'idx': ['node', 'year', ],
+            'unit': 'T$',
+        },
+        'historical_gdp': {
+            'idx': ['node', 'year', ],
+            'unit': 'T$',
+            'data_key': 'gdp0',
+        },
+        'MERtoPPP': {
+            'idx': ['node', 'year', ],
+            'unit': '-',
+        },
+        'kgdp': {
+            'idx': ['node', ],
+            'unit': '-',
+        },
+        'kpvs': {
+            'idx': ['node', ],
+            'unit': '-',
+        },
+        'depr': {
+            'idx': ['node', ],
+            'unit': '-',
+        },
+        'drate': {
+            'idx': ['node', ],
+            'unit': '-',
+        },
+        'esub': {
+            'idx': ['node', ],
+            'unit': '-',
+        },
+        'lotol': {
+            'idx': ['node', ],
+            'unit': '-',
+        },
+        'price_ref': {
+            'idx': ['node', 'sector', ],
+            'unit': 'USD/kWa',
+        },
+        'lakl': {
+            'idx': ['node', ],
+            'unit': '-',
+            'data_key': 'aconst',
+        },
+        'prfconst': {
+            'idx': ['node', 'sector', ],
+            'unit': '-',
+            'data_key': 'bconst',
+        },
+        'grow': {
+            'idx': ['node', 'year', ],
+            'unit': '-',
+            'data_key': 'growth',
+        },
+        'aeei': {
+            'idx': ['node', 'sector', 'year', ],
+            'unit': '-',
+        },
+    },
+    'vars': {
+        'DEMAND': ['node', 'commodity', 'level', 'year', 'time', ],
+        'PRICE': ['node', 'commodity', 'level', 'year', 'time', ],
+        'COST_NODAL': ['node', 'year', ],
+        'COST_NODAL_NET': ['node', 'year', ],
+        'GDP': ['node', 'year', ],
+        'I': ['node', 'year', ],
+        'C': ['node', 'year', ],
+        'K': ['node', 'year', ],
+        'KN': ['node', 'year', ],
+        'Y': ['node', 'year', ],
+        'YN': ['node', 'year', ],
+        'EC': ['node', 'year', ],
+        'UTILITY': None,
+        'PHYSENE': ['node', 'sector', 'year', ],
+        'PRODENE': ['node', 'sector', 'year', ],
+        'NEWENE': ['node', 'sector', 'year', ],
+        'grow_calibrate': ['node', 'year', ],
+        'aeei_calibrate': ['node', 'sector', 'year', ],
+    },
+    'equs': {
+        'COST_ACCOUNTING_NODAL': ['node', 'year', ],
+    },
 }
 
+MACRO_DATA_FOR_DERIVATION = {
+    'cost_ref': ['node', ],
+    'demand_ref': ['node', 'sector', ],
+}
+
+VERIFY_INPUT_DATA = [
+    'price_ref', 'lotol', 'esub', 'drate', 'depr', 'kpvs', 'kgdp',
+    'gdp_calibrate', 'aeei', 'cost_ref', 'demand_ref', 'MERtoPPP',
+]
+
+
+def _validate_data(name, df, nodes, sectors, years):
+    def validate(kind, values, df):
+        if kind not in df:
+            return
+
+        diff = set(values) - set(df[kind])
+        if diff:
```

Comment:
@gidden, this check needs to be revised as there may be some nodes or demand sectors in a scenario that are not included in MESSAGE-MACRO linkage. For example, in the global model we have a set 'cat_node', in which we define the nodes included in MACRO with the `node_type` of "economy". One solution as discussed can be to explicitly include/exclude the relevant nodes and sectors for macro-economic analysis.

Reply:
Hey @behnam2015, thanks again for the insight here. The first thing to note is that (in theory) a "clean" message model (without any prior knowledge of MACRO), should *not* have a `type_node` economy. In the test example in this PR, we add this model structure in `add_model_data()`.

As discussed in person, we agreed to add a sheet/dataframe to the input data called `config` which will have for the moment two possible columns: `ignore_node` and `ignore_sector`. Upon reading in the data, if there are entries here, they will be removed from the expected sets.

Reply:
@behnam2015, this feature is now available in the PR added as commit 7994810.

Reply:
Thanks @gidden, sounds like a very good and straightforward remedy.